sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/datastore/DatastoreV1.java

```diff
@@ -450,53 +519,136 @@ static RunQueryRequest makeRequest(Query query, @Nullable String namespace) {
-        this.query = checkNotNull(query, "query");
+      private V1Options(ValueProvider<String> project, ValueProvider<String> namespace) {
+        this.project = project;
         this.namespace = namespace;
       }
 
-      public static V1Options from(String projectId, Query query, @Nullable String namespace) {
-        return new V1Options(projectId, query, namespace);
+      public static V1Options from(String projectId, String namespace) {
+        return from(StaticValueProvider.of(projectId), StaticValueProvider.of(namespace));
       }
 
-      public Query getQuery() {
-        return query;
+      public static V1Options from(ValueProvider<String> project, ValueProvider<String> namespace) {
+        return new V1Options(project, namespace);
       }
 
       public String getProjectId() {
-        return projectId;
+        return project.get();
       }
 
       @Nullable
       public String getNamespace() {
+        return namespace == null ? null : namespace.get();
+      }
+
+      public ValueProvider<String> getProjectValueProvider() {
+        return project;
+      }
+
+      @Nullable
+      public ValueProvider<String> getNamespaceValueProvider() {
         return namespace;
       }
+
+    }
+
+    /**
+     * A DoFn that translates a Cloud Datastore gql query string to {@code Query}.
+     */
+    static class GqlQueryTranslatorFn
+        extends DoFn<ValueProvider<String>, Query> {
+      private final V1Options options;
+      private transient Datastore datastore;
+      private final V1DatastoreFactory datastoreFactory;
+
+      GqlQueryTranslatorFn(V1Options options) {
+        this(options, new V1DatastoreFactory());
+      }
+
+      @VisibleForTesting
+      GqlQueryTranslatorFn(V1Options options, V1DatastoreFactory datastoreFactory) {
+        this.options = options;
+        this.datastoreFactory = datastoreFactory;
+      }
+
+      @StartBundle
+      public void startBundle(Context c) throws Exception {
+        datastore = datastoreFactory.getDatastore(c.getPipelineOptions(), options.getProjectId());
+      }
+
+      @ProcessElement
+      public void processElement(ProcessContext c) throws Exception {
+        ValueProvider<String> gqlQuery = c.element();
+        LOG.info("User query: '{}'", gqlQuery.get());
+        // Gql query is provided, convert it to query
+        // Note: If user's query already has a limit, then that limit could take precedence
+        // over limit 0. resulting in actually reading entities but should not be a large
+        // number since service has a cap on the number of entities returned in a single request.
+        String gqlQueryWithZeroLimit = gqlQuery.get() + " limit 0";
+        GqlQuery gql = GqlQuery.newBuilder().setQueryString(gqlQueryWithZeroLimit)
+            .setAllowLiterals(true).build();
+
+        RunQueryRequest req = makeRequest(gql, options.getNamespace());
+        RunQueryResponse resp = datastore.runQuery(req);
```

Comment:
Is this because Datastore provides no API for translating a gql query to Query, except by actually running the query? This is terrible, but if it is what it is, then please document this.